numpy/_core/src/multiarray/ctors.c

```diff
@@ -1717,16 +1720,22 @@ PyArray_FromAny_int(PyObject *op, PyArray_Descr *in_descr,
         ((PyArrayObject_fields *)ret)->descr = dtype;
     }
 
-    int success = PyArray_AssignFromCache(ret, cache);
+    int succeed = PyArray_AssignFromCache(ret, cache);
 
     ((PyArrayObject_fields *)ret)->nd = out_ndim;
     ((PyArrayObject_fields *)ret)->descr = out_descr;
     Py_DECREF(dtype);
-    if (success < 0) {
+    if (succeed < 0) {
         Py_DECREF(ret);
-        return NULL;
+        goto fail;
     }
+  success:
+    NPY_END_CRITICAL_SECTION_NO_BRACKETS(array_creation_cs);
     return (PyObject *)ret;
+
+  fail:
+    NPY_END_CRITICAL_SECTION_NO_BRACKETS(array_creation_cs);
+    return NULL;
```

Comment:
Would be nice to move `Py_DECREF(dtype)` (as `XDECREF`) into the fail/success blocks (and move it down in the success one), to simplify the errors.
(Could also move a `Py_XDECREF(ret)` into the fail -- with `ret` initialized.)

Anyway, no need, just slightly improving this code (even without the change) although now we need the goto so moving cleanup into it would be good.